### ISSUE TITLE
Behavior of double dash ('--') is surprising when there are no positional argument parsers

### ISSUE THREAD

**Issue by ExpHP** [pcapriotti/optparse-applicative#211]
If a parser contains only `switches` and `strOptions` but lacks `strArguments` (or all of its `strArguments` have been parsed), then its behavior with '--' appears to be the following:

 * The first `--` is gobbled.
 * **Arguments after the first `--` continue to be parsed by the switches and strOptions**.
 * A second `--` generates an error.

I find the second point to be surprising because I would expect these arguments to be regarded as "positionals" regardless of whether there are any `strArguments` waiting to consume them.

**Demonstration:**

    import Control.Applicative
    import Options.Applicative
    
    -- Usages:
    --   main1 [-a]
    --   main2 [-a] Arg
    --   main3 [-a] [Arg...]
    data Args1 = Args1 Bool deriving Show
    data Args2 = Args2 Bool String deriving Show
    data Args3 = Args3 Bool [String] deriving Show
    
    parser1 = Args1 <$> switch (short 'a')
    parser2 = Args2 <$> switch (short 'a') <*> strArgument (metavar "Arg")
    parser3 = Args3 <$> switch (short 'a') <*> (many.strArgument) (metavar "Arg...")
    
    -- compile as main1, then subsitute 'parser1'
    --    here to compile main2 and main3
    theParser = parser1
    main = execParser (info theParser idm) >>= print

**Comparison of actual output and what I expected to see:**

    Commands:
        ./main1 -- -a
        ./main2 -- -a
        ./main3 -- -a
    Output:
        Args1 True
        Args2 False "-a"
        Args3 False ["-a"]
    Expected to see:
        Invalid argument `-a'      (mismatch)
        Args2 False "-a"           (OK)
        Args3 False ["-a"]         (OK)

    Commands:
        ./main1 -- -- -a
        ./main2 -- -- -a
        ./main3 -- -- -a
    Output:
        Invalid option `--'
        Args2 True "--"
        Args3 False ["--","-a"]
    Expected to see:
        Invalid argument `--'      (minor mismatch)
        Invalid argument `-a'      (mismatch)
        Args3 False ["--","-a"]    (OK)

**Comment by HuwCampbell** [comment]
You are correct. The behavior after `--` is that arguments are searched first, then, if none are found, options and flags are [considered](https://github.com/pcapriotti/optparse-applicative/blob/master/Options/Applicative/Common.hs#L197-L199).
The alternative is to always fail if there's no positional arguments left to find, which could be better I guess.

**Comment by ExpHP** [comment]
Ah, so it is deliberate.  My experience with it is that the first time I tried using the library, I copied an example parser that had no positionals and started playing around with it. Since the documentation mentions support for `--` and as I am fairly new to Haskell, when I encountered the behavior my first suspect was that the -- was getting eaten by some compiler magic I didn't understand (like RTS options).

I see now that the phrasing in the readme is consistent with the current behavior  ("all the following words are considered by argument parsers").  Also, try as I might, I can't come up with any example of a real situation where the current behavior causes an issue, because there is no good reason to use -- on a program with no positionals...

I guess the only "issue" then is that our mental models differ! =p
